### PR TITLE
Use regular expression matching in QuadSinkFile and prioritise longest match

### LINKED ISSUE #36: QuadSinkFiles > iriToPath should prioritize longer prefixes
Consider the following:
```json
{
  "quadSink": {
    "@type": "QuadSinkFile",
    "log": true,
    "outputFormat": "application/n-quads",
    "fileExtension": "$.nq",
    "iriToPath": {
      "http://example.org/base/": "output/base/",
      "http://example.org/": "output/other/"
    }
  }
}
```
everything will end up in `output/other`, including the IRIs starting with "http://example.org/base/".

When several prefixes apply, I would expect the longer (most specific) to apply.

### PR DESCRIPTION
**Pull request by surilindur**
This changes `QuadSinkFile` to use regular expression to string substitution mappings instead of basic prefix mapping. This should allow to, for example, map the contents of several URIs to a single file.

This PR also fixes #36 by prioritising longer regex matches.

This is a small change, but the PR looks too big because it is based off #38.

### CODE REVIEW

**Comment by rubensworks** [comment]
Let's put this on hold until #38 is merged.

**Comment by surilindur** [comment]
This is now finally ready for review! @rubensworks

File: lib/io/QuadSinkFile.ts [hunk 1]
```diff
@@ -18,19 +17,18 @@ export class QuadSinkFile implements IQuadSink {
...
     this.fileExtension = options.fileExtension;
-
     this.fileWriter = new ParallelFileWriter({ streams: 128 });
-
     this.attemptLog();
   }
 
   protected attemptLog(newLine = false): void {
     if (this.log && (this.counter % 1_000 === 0 || newLine)) {
-      readline.clearLine(process.stdout, 0);
```
Comment (rubensworks) [inline lib/io/QuadSinkFile.ts:32]: Why did you remove this?

Comment (surilindur) [reply lib/io/QuadSinkFile.ts:32]: Because the `\r` in the print just below it already does the same, so it was unnecessary.

Comment (rubensworks) [reply lib/io/QuadSinkFile.ts:32]: Theoretically it should do the same, but it's not implemented in the same way in Node AFAICS.
In any case, while I don't remember it, there's probably a reason why it was implemented like this.
So unless this causes issues, I would not touch it and keep the code as-is.

Comment (surilindur) [reply lib/io/QuadSinkFile.ts:32]: I changed it back to use the `readline` module, and removed the `\r` because the two additional function calls already achieve the same effect as that one character in practice.

Comment (rubensworks) [reply lib/io/QuadSinkFile.ts:32]: > and removed the \r because the two additional function calls already achieve the same effect as that one character in practice.

But did that break anything?
If not, let's keep it as-is, as there may be a good reason for it to be defined like that.

Comment (surilindur) [reply lib/io/QuadSinkFile.ts:32]: It did not break anything, but it also did not do anything because the two functions already did its job before it. I have added it back, and updated the unit tests again to reflect the change.

I do not think there is any good reason to have **both** the two functions calls and the carriage return. Removing the two function calls would have made the tests simpler (with less mocking and call checks), removed the imports from `readline` and reduced the amount of processing involved in actually printing the output.

Comment (rubensworks) [reply lib/io/QuadSinkFile.ts:32]: > I do not think there is any good reason to have both the two functions calls

"think" is the keyword here :-)
I agree with you that it's probably not needed, but we're not certain.
And until we're certain, we're better off with not touching it.

**Comment by rubensworks** [comment]
Thanks @surilindur!

File: lib/io/QuadSinkFile.ts (hunk 1 above)
Comment (surilindur) [reply lib/io/QuadSinkFile.ts:32]: But how can you ever be certain about something like that? Following the same logic, the newline character `\n` would need to be replaced with `EOL` from `node:os` because Windows apparently uses CRLF instead of just LF. So everywhere there is `\n` for a newline risks breakage on Windows, if someone ever uses that software on Windows. Likewise, all paths should always be joined using `join` from `node:path`, because Windows paths use `\` instead of `/`, and there are a lot of `/` in various projects to join paths manually. Everywhere there is a manual path join using `/`, there is the implicit assumption (or explicit?) that no one will ever run that thing on Windows.

I do not use Windows myself, and everywhere I have tested, `\n` works as a newline without issues. Therefore, it would have been reasonable to assume `\r` to clear the line, which it did perfectly fine during my testing. If someone would have run into issues with `\r`, they would most definitely also run into issues with `\n` as a newline.

Comment (rubensworks) [reply lib/io/QuadSinkFile.ts:32]: > But how can you ever be certain about something like that?

You can't.
As long as you're working on a closed personal projects, this is not a problem.
But we're working here with an open-source project, and we don't know where and how it's being used.
So we have to be careful with introducing breakages on other people's systems.

Comment (surilindur) [reply lib/io/QuadSinkFile.ts:32]: This was an extremely small change that, when introducing breakage, would have caused some log prints to look ugly. There was no significant risk in the change. In the unlikely scenario that someone would use this project with a terminal or other output that properly handles `\n` but cannot handle `\r` (I am pretty sure there are none), I am sure they would have opened an issue for it to let the project know.

I dislike avoiding changes that improve code readability or performance for the sake of not breaking things. There will always be someone with an obscure terminal, operating system, library or Node version that does not properly handle something very specific, but that should not be the reason to avoid doing something within an open-source project.

That is my opinion, and I dislike the fact that the change was reverted due to a reason like that. I just wanted to get this said, now I have nothing more to add to this PR. :sweat_smile:
